### PR TITLE
Setup DynamicLinks.shorten_url and DynamicLinks::ShortenedUrl.find_or_create! alternative

### PR DESCRIPTION
**Pull request by yosefbennywidyo**
Setup ShortenedUrl Subclasses:
- create current implementation ShortenedUrl on ShortenedUrlV1
- create alternative implementation ShortenedUrl on ShortenedUrlV2

Setup DynamicLinks for current and alternative shortener:
- create DynamicLinks.ShortenerV1 for current
- create DynamicLinks.ShortenerV2 for alternative
- create shorten_url_v1 and shorten_url_v2

Create steps detail to setup benchmark

### CODE REVIEW

File: lib/dynamic_links.rb [hunk 1]
```diff
@@ -50,10 +52,21 @@ def configure
     end
   end
 
-  def self.shorten_url(url, client, async: DynamicLinks.configuration.async_processing)
```
Comment (saiqulhaq) [inline lib/dynamic_links.rb:53]: kenapa method ini dihapus?
ini adalah method utama yg sering dipanggil oleh aplikasi yg pake gem ini

File: BENCHMARKS.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
+# Benchmark
```
Comment (saiqulhaq) [inline BENCHMARKS.md:1]: file benchmark ini gak perlu di commit di repo
cukup di masukin deskripsi PR nya aja

jika memang V2 lebih bagus, gak perlu bikin method v2, bisa di replace shorten yg ada, biar cukup maintain 1 jenis method shorten

dan perlu dibuat juga rspec nya
percuma jika speednya bagus tp outputnya salah

**Review by saiqulhaq — CHANGES_REQUESTED** [review]

File: lib/dynamic_links.rb (hunk 1 above)
Comment (yosefbennywidyo) [reply lib/dynamic_links.rb:53]: agar bisa membandingkan versi yang sebelum (v1) dan sesudah (v2) perubahan, akibatnya test-nya failed.
😅

**Review by saiqulhaq — COMMENTED** [review]
code class Shortener nya kok gak ada pak?
image

apa cukup nge update validasi di model nya aja biar lebih cepet?

File: test/lib/shortening_strategies/mock_strategy_test.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
+require "test_helper"
+
+class DynamicLinks::ShorteningStrategies::MockStrategyTest < ActiveSupport::TestCase
+  def setup
+    @url_shortener = DynamicLinks::ShorteningStrategies::MockStrategy.new
+  end
+
+  test "raise shorten method not implemented" do
+  	url = "https://example.com"
```
[the remark this answers is not in the document]
Comment (yosefbennywidyo) [reply test/lib/shortening_strategies/mock_strategy_test.rb:9]: Noted

[the remark this answers is not in the document]
Comment (yosefbennywidyo) [reply test/lib/shortening_strategies/mock_strategy_test.rb:9]: > indentasi

Noted

**Comment by yosefbennywidyo** [comment]
> code class Shortener nya kok gak ada pak?

File ini yang dimaksud?
Screen Shot 2024-05-29 at 22 18 17

**Comment by yosefbennywidyo** [comment]
> apa cukup nge update validasi di model nya aja biar lebih cepet?

ini salah satu alternatif, alternatif lainnya mungkin menggunakan raw SQL

**Review by saiqulhaq — COMMENTED** [review]
sorry I still don't get why by adding 
`return url_error unless url`
image

causing so much speed improvement as your benchmark result before
is it because you are testing `empty` and `exist` scenarios?

image

**Comment by yosefbennywidyo** [comment]
> sorry I still don't get why by adding `return url_error unless url` image
> 
> causing so much speed improvement as your benchmark result before is it because you are testing `empty` and `exist` scenarios?
> 
> image

My idea was to implement early exit when the url did not exist, but when I re-read the code and did a benchmark, the current code performed better.

**Pull request merged**